From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tom Levy <tomlevy93@gmail.com>
Date: Fri, 27 Mar 2020 03:56:30 +1300
Subject: Create files with correct ownership instead of using 'chown -R'

Recursive chown is slow in Docker because of the copy-on-write
filesystem. In some setups it can be extremely slow.

It is slow even when the files already have the correct owner. The
option '--from=OLD_OWNER' can be used to only modify files owned by a
particular user (if the old owner is unknown, 'find ! -user NEW_OWNER'
can be used), but it's better to create the files with the correct
ownership in the first place.
---
 docker/domserver/Dockerfile | 26 ++++++++++++++++----------
 docker/domserver/build.sh   |  2 +-
 2 files changed, 17 insertions(+), 11 deletions(-)

diff --git a/docker/domserver/Dockerfile b/docker/domserver/Dockerfile
index 74f915315..2ef9ec4be 100644
--- a/docker/domserver/Dockerfile
+++ b/docker/domserver/Dockerfile
@@ -31,17 +31,19 @@ ADD https://getcomposer.org/installer composer-setup.php
 RUN php composer-setup.php \
     && mv /composer.phar /usr/local/bin/composer
 
+RUN mkdir /domjudge-src
+
 # Install PHP dependencies
 # (also done by 'make dist', but we do it here *before* adding the rest of the source to leverage the build cache)
-COPY composer.json composer.lock /domjudge-src/domjudge/
-RUN cd /domjudge-src/domjudge && composer install --prefer-dist -o --no-scripts
+COPY --chown=domjudge composer.json composer.lock /domjudge-src/domjudge/
+RUN cd /domjudge-src/domjudge && sudo -u domjudge composer install --prefer-dist -o --no-scripts
 
 # Add DOMjudge source code and build script
-COPY . /domjudge-src/domjudge
+COPY --chown=domjudge . /domjudge-src/domjudge
 ADD docker/domserver/build.sh /domjudge-src
 
 # Bootstrap the configure/build system
-RUN make -C /domjudge-src/domjudge dist
+RUN sudo -u domjudge make -C /domjudge-src/domjudge dist
 
 # Build and install domserver
 RUN /domjudge-src/build.sh
@@ -70,20 +72,24 @@ RUN apt-get update \
 	enscript lpr \
 	&& rm -rf /var/lib/apt/lists/*
 
+# Set up user
+RUN useradd -m domjudge
+
 # Install composer
 ADD https://getcomposer.org/installer composer-setup.php
 RUN php composer-setup.php \
     && mv /composer.phar /usr/local/bin/composer
 
+RUN mkdir /opt/domjudge && chown domjudge: /opt/domjudge
+
 # Copy domserver directory and add script files
-COPY --from=domserver-build /opt/domjudge/domserver /opt/domjudge/domserver
-COPY --from=domserver-build /opt/domjudge/doc /opt/domjudge/doc
+COPY --from=domserver-build --chown=domjudge /opt/domjudge/domserver /opt/domjudge/domserver
+COPY --from=domserver-build --chown=domjudge /opt/domjudge/doc /opt/domjudge/doc
 COPY ["docker/domserver/scripts", "/scripts/"]
 
-# Add user, create PHP FPM socket dir, change permissions for domjudge directory and fix scripts
-RUN useradd -m domjudge \
-	&& mkdir -p /run/php \
-	&& chown -R domjudge: /opt/domjudge \
+# Create PHP FPM socket dir, change ownership of some domjudge directories and fix scripts
+# ("chown -R domjudge: /opt/domjudge" not required; verify using "chown --changes ...")
+RUN mkdir -p /run/php \
 	&& chown -R www-data: /opt/domjudge/domserver/tmp \
 	# for DOMjudge <= 7.2 (submitdir was removed in commit DOMjudge/domjudge@d66725038)
 	&& if [ -d /opt/domjudge/domserver/submissions ]; then chown -R www-data: /opt/domjudge/domserver/submissions; fi \
diff --git a/docker/domserver/build.sh b/docker/domserver/build.sh
index f616c9e7b..82270a551 100755
--- a/docker/domserver/build.sh
+++ b/docker/domserver/build.sh
@@ -1,7 +1,7 @@
 #!/bin/bash -e
 
 cd /domjudge-src/domjudge*
-chown -R domjudge: .
+# ("chown -R domjudge: ." not required; verify using "chown --changes ...")
 sudo -u domjudge ./configure -with-baseurl=http://localhost/
 
 # Passwords should not be included in the built image. We create empty files here to prevent passwords from being generated.
-- 
2.20.1

